src/runtime/particle-execution-context.ts

```diff
@@ -185,29 +187,56 @@ export class ParticleExecutionContext implements StorageCommunicationEndpointPro
           }
         }
         message.id = id;
-
-        // The presence of an entity id indicates the message should be sent to a Backing store. Proxy messages sent to
-        // Backing stores requires an entity id in order to redirect the message to the correct store.
-        if (entityId != null) {
-          return new Promise(resolve =>
-            pec.apiPort.BackingProxyMessage(storageProxy, message, entityId, ret => resolve(ret)));
+        return  new Promise((resolve) =>
+            pec.apiPort.ProxyMessage(storageProxy, message, ret => resolve(ret)));
+        },
+
+        setCallback(callback: ProxyCallback<CRDTTypeRecord>): void {
+          idPromise = new Promise<number>((resolve) =>
+          pec.apiPort.Register(storageProxy, callback, retId => resolve(retId)));
+        },
+        reportExceptionInHost(exception: PropagatedException): void {
+          pec.apiPort.ReportExceptionInHost(exception);
+        },
+        getChannelConstructor(): ChannelConstructor {
+          return pec;
         }
-
-        return new Promise(resolve =>
-          pec.apiPort.ProxyMessage(storageProxy, message, ret => resolve(ret)));
-      },
-
-      setCallback(callback: ProxyCallback<CRDTTypeRecord>): void {
-        idPromise = new Promise<number>((resolve) =>
-          pec.apiPort.Register(storageProxy, x => storageProxy.onMessage(x), retId => resolve(retId)));
-      },
-      reportExceptionInHost(exception: PropagatedException): void {
-        pec.apiPort.ReportExceptionInHost(exception);
-      },
-      getChannelConstructor(): ChannelConstructor {
-        return pec;
-      }
-    };
+      };
+    } else if (storageProxy instanceof BackingStorageProxy) {
+      return {
+        async onProxyMessage(message: ProxyMessage<CRDTTypeRecord>): Promise<boolean> {
+          if (idPromise == null) {
+            throw new Error('onProxyMessage called without first calling setCallback!');
+          }
+          if (id == null) {
+            id = await idPromise;
+            if (id == null) {
+              throw new Error('undefined id received .. somehow');
+            }
+          }
+          message.id = id;
```

Comment:
PTAL. With the suggested simplification, ```id``` was never set to the resolved ```idPromise```. So I got rid of ```id``` altogether. I do think this simplification is a good idea because it's easier to follow the control flow.

Reply:
Yes, I missed the fact that `id` was a captured outer variable. So this new approach will work and is generally more elegant, with one caveat: since you're now always awaiting, there will always be an async bump on each call to `onProxyMessage` (even when `idPromise` has already been resolved). The old way only had the bump on the first proxy message.

I think the trivial runtime inefficiency caused by that is worth it, since the new way is quite a lot cleaner.

Reply:
Agreed :)